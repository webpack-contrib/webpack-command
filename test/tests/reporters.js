const strip = require('strip-ansi');

const parse = require('../../lib/reporters/parse');
const Reporter = require('../../lib/reporters/Reporter');
const { apply, build, test, validate } = require('../util');

test('reporter parse util', module, () => {
  it('should parse hidden', () => {
    const stats = {
      filteredAssets: 1,
      filteredModules: 2,
    };
    const result = parse.hidden(stats);

    expect(result).toMatchSnapshot();
  });

  it('should parse status: cacheable', () => {
    const result = parse.status({ cacheable: false });

    expect(result).toMatchSnapshot();
  });

  it('should parse status: optional', () => {
    const result = parse.status({ optional: true });

    expect(result).toMatchSnapshot();
  });

  it('should parse status: prefetched', () => {
    const result = parse.status({ prefetched: true });

    expect(result).toMatchSnapshot();
  });
});

test('Reporter', module, () => {
  it('should have methods', () => {
    expect(Reporter).toBeDefined();

    const reporter = new Reporter({ compiler: {}, config: {} });

    expect(reporter).toBeInstanceOf(Reporter);
    expect(reporter.compiler).toBeDefined();
    expect(reporter.config).toBeDefined();
    expect(reporter.progress).toBeDefined();
    expect(reporter.render).toBeDefined();

    expect(reporter.progress()).toBeUndefined();
    expect(reporter.render()).toBeUndefined();
  });
});

test('StylishReporter', module, () => {
  for (const name of [
    'stylish',
    'stylish-multi',
    'problems/stylish-problems',
  ]) {
    const fixture = `stylish/${name}`;
    const fixtureType = 'reporters';
    const opts = { fixture, fixtureType };

    let config;

    it(`${name}: reporter should validate`, () => {
      expect(validate(opts)).toEqual(true);
    });

    it(`${name}: reporter should apply`, () => {
      config = apply(opts);

      expect(config).toMatchSnapshot();
    });

    it(`${name}: reporter should build`, () =>
      build(config).then((result) => {
        expect(
          strip(result)
            .replace(/Δt \d+ms/g, '<duration>')
            .replace(/\d+\.\d+ (KiB|kB)/g, '<size>')
            .replace(/[a-f0-9]{20}/g, '<hash>')
        ).toMatchSnapshot();
      }));
  }
});

test('JsonReporter', module, () => {
  for (const name of ['json', 'json-multi', 'problems/problems']) {
    const fixture = `json/${name}`;
    const fixtureType = 'reporters';
    const opts = { fixture, fixtureType };

    let config;

    it(`${name}: reporter should validate`, () => {
      expect(validate(opts)).toEqual(true);
    });

    it(`${name}: reporter should apply`, () => {
      config = apply(opts);

      expect(config).toMatchSnapshot();
    });

    it(`${name}: reporter should build`, () =>
      build(config).then((result) => {
        expect(result.length).toBeGreaterThan(0);

        const json = JSON.parse(result);

        expect(Object.keys(json).length).toBeGreaterThan(0);

        if (name === 'problems/problems') {
          expect(json.errors.length).toBeGreaterThan(0);
          expect(json.warnings.length).toBeGreaterThan(0);
        }
      }));
  }
});

test('reporters + bail', module, () => {
  for (const name of ['Basic', 'Json', 'Stylish']) {
    it(`${name}: render should return error when bail set`, () => {
      const reporterPath = `../../lib/reporters/${name}Reporter`;
      // eslint-disable-next-line global-require, import/no-dynamic-require
      const ReporterClass = require(reporterPath);
      const reporter = new ReporterClass({});
      const error = new Error('test');
      expect(reporter.render(error, null)).toMatchSnapshot();
    });
  }
});

test('bad reporter name', module, () => {
  const fixture = `json/json`;
  const fixtureType = 'reporters';
  const opts = { fixture, fixtureType };

  let config;

  it(`should apply`, () => {
    config = apply(opts);
    config.reporter = 'batman';
    expect(config).toMatchSnapshot();
  });

  it(`should default to stylish`, () =>
    build(config).then((result) => {
      expect(result.length).toBeGreaterThan(0);
      expect(
        strip(result)
          .replace(/Δt \d+ms/g, '<duration>')
          .replace(/\d+\.\d+ (KiB|kB)/g, '<size>')
          .replace(/[a-f0-9]{20}/g, '<hash>')
      ).toMatchSnapshot();
    }));
});
