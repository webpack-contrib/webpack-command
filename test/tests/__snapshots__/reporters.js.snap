// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonReporter > json: reporter should apply #0 1`] = `
Object {
  "entry": "<PROJECT_ROOT>/test/fixtures/common/entry-a.js",
  "mode": "development",
  "plugins": Array [
    NamedModulesPlugin {
      "options": Object {},
    },
  ],
  "reporter": "json",
}
`;

exports[`JsonReporter > json-multi: reporter should apply #0 1`] = `
Array [
  Object {
    "entry": "<PROJECT_ROOT>/test/fixtures/common/entry-a.js",
    "mode": "development",
    "plugins": Array [
      NamedModulesPlugin {
        "options": Object {},
      },
    ],
    "reporter": "json",
  },
  Object {
    "entry": Array [
      "<PROJECT_ROOT>/test/fixtures/common/entry-b.js",
      "<PROJECT_ROOT>/test/fixtures/common/entry-c.js",
    ],
    "mode": "development",
    "plugins": Array [
      NamedModulesPlugin {
        "options": Object {},
      },
    ],
    "reporter": "json",
  },
]
`;

exports[`JsonReporter > problems/problems: reporter should apply #0 1`] = `
Object {
  "entry": "<PROJECT_ROOT>/test/fixtures/reporters/json/problems/entry-problems.js",
  "mode": "development",
  "plugins": Array [
    NamedModulesPlugin {
      "options": Object {},
    },
  ],
  "reporter": "json",
}
`;

exports[`StylishReporter > problems/stylish-problems reporter should apply #0 1`] = `
Object {
  "entry": "<PROJECT_ROOT>/test/fixtures/reporters/stylish/problems/entry-problems.js",
  "mode": "development",
  "plugins": Array [
    NamedModulesPlugin {
      "options": Object {},
    },
  ],
}
`;

exports[`StylishReporter > problems/stylish-problems reporter should build #0 1`] = `
"
webpack v4.6.0

<hash>
  size     name                     module                                                        status
  160 B    problems sync recursive  ./test/fixtures/reporters/stylish/problems sync               built
  148 B    entry-problems.js        ./test/fixtures/reporters/stylish/problems/entry-problems.js  built ⚠ ✖
  177 B    image.jpg                ./test/fixtures/reporters/stylish/problems/image.jpg          built failed ✖

  size     name                     asset                                                         status
  <size>  main                     main.js                                                       emitted

  <duration> 


./test/fixtures/reporters/stylish/problems/entry-problems.js
  5:15  error    Module not found: Error: Can't resolve 'test' in
                 '<PROJECT_ROOT>/test/fixtures/reporters/stylish/problems'
  2:12  warning  Critical dependency: require function is used in a way in which
                 dependencies cannot be statically extracted

./test/fixtures/reporters/stylish/problems/image.jpg
  1:0  error  Module parse failed: Unexpected character '�' You may need an
              appropriate loader to handle this file type. (Source code omitted for
              this binary file)  @
              ./test/fixtures/reporters/stylish/problems/entry-problems.js 7:0-22

✖ 3 problems (2 errors, 1 warning)"
`;

exports[`StylishReporter > problems/stylish-problems: reporter should apply #0 1`] = `
Object {
  "entry": "<PROJECT_ROOT>/test/fixtures/reporters/stylish/problems/entry-problems.js",
  "mode": "development",
  "plugins": Array [
    NamedModulesPlugin {
      "options": Object {},
    },
  ],
}
`;

exports[`StylishReporter > problems/stylish-problems: reporter should build #0 1`] = `
"
webpack v4.14.0

<hash>
  size   name                     module                                                        status
  160 B  problems sync recursive  ./test/fixtures/reporters/stylish/problems sync               built
  148 B  entry-problems.js        ./test/fixtures/reporters/stylish/problems/entry-problems.js  built ✖
  177 B  image.jpg                ./test/fixtures/reporters/stylish/problems/image.jpg          built failed ✖

  size   name                     asset                                                         status
  6 kB   main                     main.js                                                       emitted

  <duration> 


./test/fixtures/reporters/stylish/problems/entry-problems.js
  5:15  error  Module not found: Error: Can't resolve 'test' in
               '<PROJECT_ROOT>/test/fixtures/reporters/stylish/problems'

./test/fixtures/reporters/stylish/problems/image.jpg 1:0
  1:0  error  Module parse failed: Unexpected character '�' You may need an
              appropriate loader to handle this file type. (Source code omitted for
              this binary file)  @
              ./test/fixtures/reporters/stylish/problems/entry-problems.js 7:0-22

./test/fixtures/reporters/stylish/problems/entry-problems.js 2:12-19
  0:0  warning  Critical dependency: require function is used in a way in which
                dependencies cannot be statically extracted

✖ 3 problems (2 errors, 1 warning)"
`;

exports[`StylishReporter > stylish reporter should apply #0 1`] = `
Object {
  "entry": "<PROJECT_ROOT>/test/fixtures/common/entry-a.js",
  "mode": "development",
  "plugins": Array [
    NamedModulesPlugin {
      "options": Object {},
    },
  ],
}
`;

exports[`StylishReporter > stylish reporter should build #0 1`] = `
"
webpack v4.6.0

<hash>
  size     name        module                             status
  41 B     entry-a.js  ./test/fixtures/common/entry-a.js  built

  size     name        asset                              status
  <size>  main        main.js                            emitted

  <duration> 
"
`;

exports[`StylishReporter > stylish: reporter should apply #0 1`] = `
Object {
  "entry": "<PROJECT_ROOT>/test/fixtures/common/entry-a.js",
  "mode": "development",
  "plugins": Array [
    NamedModulesPlugin {
      "options": Object {},
    },
  ],
}
`;

exports[`StylishReporter > stylish: reporter should build #0 1`] = `
"
webpack v4.14.0

<hash>
  size     name        module                             status
  41 B     entry-a.js  ./test/fixtures/common/entry-a.js  built

  size     name        asset                              status
  <size>  main        main.js                            emitted

  <duration> 
"
`;

exports[`StylishReporter > stylish-multi reporter should apply #0 1`] = `
Array [
  Object {
    "entry": "<PROJECT_ROOT>/test/fixtures/common/entry-a.js",
    "mode": "development",
    "plugins": Array [
      NamedModulesPlugin {
        "options": Object {},
      },
    ],
  },
  Object {
    "entry": Array [
      "<PROJECT_ROOT>/test/fixtures/common/entry-b.js",
      "<PROJECT_ROOT>/test/fixtures/common/entry-c.js",
    ],
    "mode": "development",
    "plugins": Array [
      NamedModulesPlugin {
        "options": Object {},
      },
    ],
  },
]
`;

exports[`StylishReporter > stylish-multi reporter should build #0 1`] = `
"
webpack v4.6.0

<hash>
  size     name        module                             status
  41 B     entry-a.js  ./test/fixtures/common/entry-a.js  built

  size     name        asset                              status
  <size>  main        main.js                            emitted

  <duration> 

<hash>
  size     name        module                                                                     status
  41 B     entry-b.js  ./test/fixtures/common/entry-b.js                                          built
  41 B     entry-c.js  ./test/fixtures/common/entry-c.js                                          built
  40 B     0           multi ./test/fixtures/common/entry-b.js ./test/fixtures/common/entry-c.js  built

  size     name        asset                                                                      status
  <size>  main        main.js                                                                    emitted

  <duration> 

total <duration>"
`;

exports[`StylishReporter > stylish-multi: reporter should apply #0 1`] = `
Array [
  Object {
    "entry": "<PROJECT_ROOT>/test/fixtures/common/entry-a.js",
    "mode": "development",
    "plugins": Array [
      NamedModulesPlugin {
        "options": Object {},
      },
    ],
  },
  Object {
    "entry": Array [
      "<PROJECT_ROOT>/test/fixtures/common/entry-b.js",
      "<PROJECT_ROOT>/test/fixtures/common/entry-c.js",
    ],
    "mode": "development",
    "plugins": Array [
      NamedModulesPlugin {
        "options": Object {},
      },
    ],
  },
]
`;

exports[`StylishReporter > stylish-multi: reporter should build #0 1`] = `
"
webpack v4.14.0

<hash>
  size     name        module                             status
  41 B     entry-a.js  ./test/fixtures/common/entry-a.js  built

  size     name        asset                              status
  <size>  main        main.js                            emitted

  <duration> 

<hash>
  size     name        module                                                                     status
  41 B     entry-b.js  ./test/fixtures/common/entry-b.js                                          built
  41 B     entry-c.js  ./test/fixtures/common/entry-c.js                                          built
  40 B     0           multi ./test/fixtures/common/entry-b.js ./test/fixtures/common/entry-c.js  built

  size     name        asset                                                                      status
  <size>  main        main.js                                                                    emitted

  <duration> 

total <duration>"
`;

exports[`bad reporter name > should apply #0 1`] = `
Object {
  "entry": "<PROJECT_ROOT>/test/fixtures/common/entry-a.js",
  "mode": "development",
  "plugins": Array [
    NamedModulesPlugin {
      "options": Object {},
    },
  ],
  "reporter": "batman",
}
`;

exports[`bad reporter name > should default to stylish #0 1`] = `
"
webpack v4.14.0

<hash>
  size     name        module                             status
  41 B     entry-a.js  ./test/fixtures/common/entry-a.js  built

  size     name        asset                              status
  <size>  main        main.js                            emitted

  <duration> 
"
`;

exports[`reporter parse util > should parse hidden #0 1`] = `"(1 asset, 2 modules hidden)"`;

exports[`reporter parse util > should parse status: cacheable #0 1`] = `
Array [
  "no-cache",
]
`;

exports[`reporter parse util > should parse status: optional #0 1`] = `
Array [
  "optional",
]
`;

exports[`reporter parse util > should parse status: prefetched #0 1`] = `
Array [
  "prefetch",
]
`;

exports[`reporters + bail > Basic: render should return error when bail set #0 1`] = `[Error: test]`;

exports[`reporters + bail > Basic: should return error when bail set #0 1`] = `[Error: test]`;

exports[`reporters + bail > Json: render should return error when bail set #0 1`] = `[Error: test]`;

exports[`reporters + bail > Json: should return error when bail set #0 1`] = `[Error: test]`;

exports[`reporters + bail > Stylish: render should return error when bail set #0 1`] = `[Error: test]`;

exports[`reporters + bail > Stylish: should return error when bail set #0 1`] = `[Error: test]`;
